File: src/OrganizationReport.ts
import { UsageItem } from './UsageItem';

type Usage = { 
	netAmount: number
	grossAmount: number
	discountAmount: number
};
type OrgUsage = Usage & {
	organization: string;
}

type OrganizationReport = { 
	usageByOrg: OrgUsage[], sumUsage: Usage 
	usageItems: UsageItem[]
};

function createOrganizationReport(usageItems: UsageItem[]): OrganizationReport {
	let overallUsage: Usage = {
		netAmount: 0,
		grossAmount: 0,
		discountAmount: 0
	};

	const sumByOrgs = usageItems.reduce((acc, item) => {
		if (!acc[item.organizationName]) {
			acc[item.organizationName] = { 
				netAmount: 0, 
				grossAmount: 0, 
				discountAmount: 0
			};
		}
		acc[item.organizationName].netAmount += item.netAmount;
		acc[item.organizationName].grossAmount += item.grossAmount;
		acc[item.organizationName].discountAmount += item.discountAmount;

		overallUsage.netAmount += item.netAmount;
		overallUsage.grossAmount += item.grossAmount;
		overallUsage.discountAmount += item.discountAmount;

		return acc;
	}, {} as { [key: string]: Usage });

	const orgUsage: OrgUsage[] = Object.entries(sumByOrgs)
		.map(([organization, usage]) => ({
			organization,
			...usage
		})).sort((a, b) => b.netAmount - a.netAmount);

	return {
		usageByOrg: orgUsage,
		sumUsage: overallUsage,
		usageItems
	};
}

export {
	createOrganizationReport
};

File: organization_report.go
package main

import "sort"

type Usage struct {
	NetAmount      float64
	GrossAmount    float64
	DiscountAmount float64
}

type OrgUsage struct {
	Usage
	Organization string
}

type OrganizationReport struct {
	UsageByOrg []OrgUsage
	SumUsage   Usage
	UsageItems []UsageItem
}

func NewOrganizationReport(usageItems []UsageItem) OrganizationReport {
	overallUsage := Usage{}

	sumByOrgs := make(map[string]Usage)

	for _, item := range usageItems {
		usage := sumByOrgs[item.OrganizationName]
		usage.NetAmount += item.NetAmount
		usage.GrossAmount += item.GrossAmount
		usage.DiscountAmount += item.DiscountAmount
		sumByOrgs[item.OrganizationName] = usage

		overallUsage.NetAmount += item.NetAmount
		overallUsage.GrossAmount += item.GrossAmount
		overallUsage.DiscountAmount += item.DiscountAmount
	}

	usageByOrg := make([]OrgUsage, 0, len(sumByOrgs))
	for org, usage := range sumByOrgs {
		usageByOrg = append(usageByOrg, OrgUsage{Usage: usage, Organization: org})
	}

	sort.Slice(usageByOrg, func(i, j int) bool {
		return usageByOrg[j].NetAmount < usageByOrg[i].NetAmount
	})

	return OrganizationReport{
		UsageByOrg: usageByOrg,
		SumUsage:   overallUsage,
		UsageItems: usageItems,
	}
}
